README
Group-Git

Tool for operating on many git repositories at once. The repositories can be
managed by Github, Bitbucket, Gitosis or not at all. If managed uncloned
repositories will be automatically cloned.

INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Group::Git

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Group-Git

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Group-Git

    CPAN Ratings
        http://cpanratings.perl.org/d/Group-Git

    Search CPAN
        http://search.cpan.org/dist/Group-Git/

    Source Code
        git://github.com/ivanwills/Group-Git.git

COPYRIGHT AND LICENCE

Copyright (C) 2013 Ivan Wills

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Group-Git

0.6.2   2017-06-26
        Updating for release of version 0.6.2
        Added in new lib to config (Ivan Wills)
        Fixed usage documentation (Ivan Wills)

0.6.1   2017-06-18
        Updating for release of version 0.6.1
        Adding new file for testing (Ivan Wills)
        Added new file (Ivan Wills)
        Adding some documentation (Ivan Wills)
        New command to give a brief status of each repository (Ivan Wills)
        Fixing removing the checked thread (Ivan Wills)
        Changing which thread in the array is checked to see if it's running (Ivan Wills)
        Some extra logic for coping with very large numbers of repositories to process while threadding (Ivan Wills)

0.6.0   2017-02-26
        Updating for release of version 0.6.0
        Updated dates (Ivan Wills)
        Making things quieter for repos with no remotes set (Ivan Wills)
        Fixed issue with spaces in directory names (Ivan Wills)

0.5.10   2016-12-12
        Updating for release of version 0.5.10
        Fixing not rapping tag hash (Ivan Wills)
        Adding a default tags (Ivan Wills)
        Using the project object to get tags now (Ivan Wills)
        Added saving tags to config (Ivan Wills)
        Using sorted tag for repo determination (Ivan Wills)
        Adding recording FS tags along with a repo object (Ivan Wills)
        Added perl 5.24 to travis build (Ivan Wills)

0.5.9   2016-08-28
        Updating for release of version 0.5.9
        Upped minimum perl version (Ivan Wills)
        Fixed spelling (Ivan Wills)
        Cleaned up documentation a bit (Ivan Wills)
        Better description (Ivan Wills)
        Fixed references to README (Ivan Wills)
        Fixed code blocks (Ivan Wills)
        Converted syntax (Ivan Wills)
        Converting to Markdown due to githubs woeful POD support (Ivan Wills)
        More documentation (Ivan Wills)
        Fied date (Ivan Wills)
        Adding lots of documentation (Ivan Wills)
        Added documentation of --max (Ivan Wills)
        Fixed issue with large number of repositories causing exessive number of threads (Ivan Wills)
        Added setting the directory when running each thread to stop leaking of directory changes (Ivan Wills)
        Added threadding output (Ivan Wills)
        Added separate option (Ivan Wills)
        Adding README to documentation to keep up to date (Ivan Wills)

0.5.8   2016-08-21
        Updating for release of version 0.5.8
        Updated date (Ivan Wills)
        Updated config (Ivan Wills)
        Synced manifest (Ivan Wills)
        Added reminder (Ivan Wills)
        Cleaned up if statement (Ivan Wills)
        Worked on pull's documentation (Ivan Wills)
        Dropped travis builds as it is using an old version of Perl 5.8 (Ivan Wills)
        VTide and editor configs (Ivan Wills)

0.5.7   2015-10-23
        Updating for release of version 0.5.7
        Added comment about mech property (Ivan Wills)
        Updated other repo type documentation (Ivan Wills)
        Improving on the comments for Stash (Ivan Wills)
        Added missing mech object (Ivan Wills)

0.5.6   2015-09-08
        Updating for release of version 0.5.6
        Fixed the pull/clone logic (Ivan Wills)
        Fixed using file rather than path (Ivan Wills)

0.5.5   2015-08-28
        Updating for release of version 0.5.5
        Removed creation of makefile (Ivan Wills)

0.5.4   2015-07-19
        Updating for release of version 0.5.4
        Converted to Path::Tiny (Ivan Wills)
        Moved caching out of Stash to Repo so Bitbucket can use it (Ivan Wills)
        Merge remote-tracking branch 'origin/cacheable' (Ivan Wills)

0.5.3   2015-07-06
        Updating for release of version 0.5.3
        Added setting an environment variable GROUP_GIT_NAME with the current repos name (Ivan Wills)
        Added Perl 5.22 to list of perl versions to build with (Ivan Wills)

0.5.2   2015-05-30
        Updating for release of version 0.5.2

0.5.1   2015-05-30
        Updating for release of version 0.5.1
        Added missing dependencies (Ivan Wills)
        Added missing documentation (Ivan Wills)
        Fixed the itterating over Stashes pages of results (Ivan Wills)
        No longer need the tag (Ivan Wills)
        Added verbose mode to show the count of found repositories (Ivan Wills)
        Fixed loading Taggers (Ivan Wills)
        Moved loading of "tagger" modules to Module::Plugin (Ivan Wills)
        Removed previously added suggestion (Ivan Wills)
        Converted to use Module::Pluggable for loading commands (Ivan Wills)
        Added reminder to try to use Module::Plugable (Ivan Wills)

0.5.0   2015-04-06
        Updating for release of version 0.5.0
        Added missing modules to tests (Ivan Wills)
        Fixed spelling (Ivan Wills)
        Fixed up dependencies (Ivan Wills)
        Moved from Moose to Moo (Ivan Wills)
        Cleaned up documentation a little (Ivan Wills)
        Fixed config file reference (Ivan Wills)
        More docs (Ivan Wills)
        Added more documentation about tags (Ivan Wills)
        Added documentation (Ivan Wills)
        Fixed issue when there is no tagging module available (Ivan Wills)
        Fixed manifest files (Ivan Wills)
        Removed warn statement (Ivan Wills)
        Added auto tagging Classes (Ivan Wills)
        Fixed require used file name (Ivan Wills)
        Implementing tagging base class (Ivan Wills)
        Added specific Test::More version for the features being used (Ivan Wills)
        Changed log to better fit Getopt::Alt's capabilities (Ivan Wills)
        Added using newer version of Getopt::Alt for more flexible arguments (to better match git) (Ivan Wills)
        Added coveralls coverage badge (Ivan Wills)
        Added more complete list of pull options (Ivan Wills)
        Added config for coveralls.io (Ivan Wills)
        Made fetch quiet (Ivan Wills)
        Added documentation about pull and sh (Ivan Wills)

lib/Group/Git/Cmd/Pull.pm
package Group::Git::Cmd::Pull;

# Created on: 2013-05-06 21:57:07
# Create by:  Ivan Wills
# $Id$
# $Revision$, $HeadURL$, $Date$
# $Revision$, $Source$, $Date$

use Moo::Role;
use strict;
use warnings;
use version;
use Carp;
use English qw/ -no_match_vars /;
use File::chdir;
use Getopt::Alt;

our $VERSION = version->new('0.6.2');

requires 'repos';
requires 'verbose';

my $opt = Getopt::Alt->new(
    { help => __PACKAGE__, },
    [
        'quiet|q!',
        'verbose|v',
        #'recurse-submodules=s![=yes|on-demand|no',
        'recurse-submodules=[yes|on-demand|no]',
        'commit!',
        'edit|e!',
        'ff!',
        'ff-only',
        #'log=d!',
        'log=d',
        'stat|n!',
        'squash!',
        'strategy|s=s',
        'strategy-option|X=s',
        'verify-signatures!',
        'summary!',
        #'rebase|r=s![=false|true|preserve]',
        'rebase|r=[false|true|preserve]',
        'all',
        'append|a',
        'depth=i',
        'unshallow',
        'update-shallow',
        'force|f',
        'keep|k',
        'no-tags',
        'update-head-ok|u',
        'upload-pack=s',
        'progress',
    ]
);

sub update_start { shift->pull_start($_[0], 'update') }
sub pull_start {
    $opt->process;
    return;
}

sub update { shift->pull($_[0], 'update') }
sub pull {
    my ($self, $name, $type) = @_;
    $type ||= 'pull';

    my $repo = $self->repos->{$name};
    my $cmd;
    my $dir;

    if ( -d $name ) {
        {
            local $CWD = $name;

            # check that there is a remote
            my $remotes = `git remote 2> /dev/null`;
            chomp $remotes;

            if ( ! $remotes && ! $self->verbose ) {
                return;
            }
        }

        $dir = $name;
        my @args = map {
                $opt->opt->{$_} eq '0'   ? "--no-$_"
                : $opt->opt->{$_} eq '1' ? "--$_"
                :                          "--$_=" . $opt->opt->{$_};
            }
            keys %{ $opt->opt };
        $cmd = join ' ', 'git', map { $self->shell_quote } $type, @args, @ARGV;
    }
    elsif ( $repo->git ) {
        $cmd = join ' ', 'git', 'clone', map { $self->shell_quote } $repo->git, $name;
    }
    else {
        return;
    }

    local $CWD = $dir if $dir;
    warn "$cmd\n" if $self->verbose > 1;
    return `$cmd 2>&1` if !$opt->opt->quiet;

    my @ans = `$cmd 2>&1`;

    return if @ans == 1 && $ans[0] =~ /^Already \s up-to-date[.]$/xms;

    return wantarray ? @ans : join '', @ans;
}

1;

__END__

=head1 NAME

Group::Git::Cmd::Pull - Pull latest versions of all repositories or clone any that are missing

=head1 VERSION

This documentation refers to Group::Git::Cmd::Pull version 0.6.2.

=head1 SYNOPSIS

    group-git pull (options)

  OPTIONS:
    -q --quiet
       --no-quiet
    -v --verbose
       --recurse-submodules=[yes|on-demand|no]
       --commit
       --no-commit
    -e --edit
       --no-edit
       --ff
       --no-ff
       --ff-only
       --log=d
    -n --stat
       --no-stat
       --squash
       --no-squash
    -s --strategy[=]str
    -X --strategy-option[=]str
       --verify-signatures
       --no-verify-signatures
       --summary
       --no-summary
    -r --rebase[=][false|true|preserve]
       --all
    -a --append
       --depth=i
       --unshallow
       --update-shallow
    -f --force
    -k --keep
       --no-tags
    -u --update-head-ok
       --upload-pack=s
       --progress

=head1 DESCRIPTION

=head1 SUBROUTINES/METHODS

=over 4

=item C<pull ($name[, 'update'])>

Runs git pull on all repositories, if a repository doesn't exist on disk this
will clone that repository.

=item C<update ($name)>

Runs git update on all repositories, if a repository doesn't exist on disk this
will clone that repository.

=item C<pull_start ()>

Pre-process pull options

=item C<update_start ()>

Pre-process update options

=back

=head1 DIAGNOSTICS

=head1 CONFIGURATION AND ENVIRONMENT

=head1 DEPENDENCIES

=head1 INCOMPATIBILITIES

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.

Please report problems to Ivan Wills (ivan.wills@gmail.com).

Patches are welcome.

=head1 AUTHOR

Ivan Wills - (ivan.wills@gmail.com)

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2013 Ivan Wills (14 Mullion Close, Hornsby Heights, NSW Australia 2077).
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.  This program is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut
